code/go/0chain.net/smartcontract/minersc/collect_reward.go

```
@@ -20,87 +20,49 @@ func (ssc *MinerSmartContract) collectReward(
-	} else {
-		providers = []string{prr.ProviderId}
-	}
-
-	var totalMinted currency.Coin
-	for _, providerID := range providers {
-		var provider *MinerNode
-		switch prr.ProviderType {
-		case spenum.Miner:
-			provider, err = getMinerNode(providerID, balances)
-		case spenum.Sharder:
-			provider, err = ssc.getSharderNode(providerID, balances)
-		default:
-			err = fmt.Errorf("unsupported provider type %s", prr.ProviderType.String())
-		}
-		if err != nil {
-			return "", common.NewError("collect_reward_failed", err.Error())
-		}
+	minted, err := stakepool.CollectReward(
+		input,
+		func(
+			crr stakepool.CollectRewardRequest, balances cstate.StateContextI,
+		) (currency.Coin, error) {
+			var provider *MinerNode
+			var err error
+			switch crr.ProviderType {
+			case spenum.Miner:
+				provider, err = getMinerNode(crr.ProviderId, balances)
+			case spenum.Sharder:
+				provider, err = getSharderNode(crr.ProviderId, balances)
+			default:
+				err = fmt.Errorf("unsupported provider type %s", crr.ProviderType.String())
+			}
+			if err != nil {
+				return 0, err
+			}
 
-		_, ok := provider.Pools[txn.ClientID]
-		if ok || providerID == txn.ClientID ||
-			provider.Settings.DelegateWallet == txn.ClientID {
 			minted, err := provider.StakePool.MintRewards(
-				txn.ClientID, providerID, prr.ProviderType, usp, balances)
+				txn.ClientID, crr.ProviderId, crr.ProviderType, balances)
 			if err != nil {
-				return "", common.NewErrorf("collect_reward_failed",
-					"error emptying account, %v", err)
+				return 0, err
 			}
 
 			if err := provider.save(balances); err != nil {
-				return "", common.NewErrorf("collect_reward_failed",
-					"error saving stake pool, %v", err)
+				return 0, err
 			}
 
-			tm, err := currency.AddCoin(totalMinted, minted)
-			if err != nil {
-				return "", common.NewErrorf("collect_reward_failed", "error adding total minted token: %v", err)
-			}
-
-			totalMinted = tm
+			return minted, nil
+		},
+		balances,
+	)
```

We should always check error first to avoid 'node not found' error getting ignored and then cause 'state hash mismatch' error. 

Imagine the miner got 'node not found' error here, all the state changes above would be invalid, but if we continue and somehow get 'value not present' or some other chargerable error below in line 62, this txn will be included to the block, and due to the invalid state we have got above, this block will get different from other miners. Although this will be finally captured on top txn executor, but we'd better avoid this kind of error at the beginning.